kv/kvserver: TestReportUnreachableRemoveRace failed

[(kv/kvserver).TestReportUnreachableRemoveRace failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2599972&tab=buildLog) on [master@dd1976daa787371fc61d7fb382c3c0744141ac4b](https://github.com/cockroachdb/cockroach/commits/dd1976daa787371fc61d7fb382c3c0744141ac4b):

```
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).ChangeReplicas
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_command.go:964
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeAdminBatch
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:584
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendWithRangeID
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:100
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:36
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:195
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).Send
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/stores.go:177
          | github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal.func1
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:874
          | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:326
          | github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:862
          | github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:900
          | github.com/cockroachdb/cockroach/pkg/roachpb._Internal_Batch_Handler.func1
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/api.pb.go:9083
          | github.com/cockroachdb/cockroach/pkg/util/tracing.ServerInterceptor.func1
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/grpc_interceptor.go:124
          | google.golang.org/grpc.getChainUnaryHandler.func1
          | 	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:921
          | github.com/cockroachdb/cockroach/pkg/rpc.NewServer.func1
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:176
          | google.golang.org/grpc.getChainUnaryHandler.func1
          | 	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:921
          | github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:60
          | google.golang.org/grpc.chainUnaryServerInterceptors.func1
          | 	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:907
          | github.com/cockroachdb/cockroach/pkg/roachpb._Internal_Batch_Handler
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/api.pb.go:9085
          | google.golang.org/grpc.(*Server).processUnaryRPC
          | 	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1082
          | google.golang.org/grpc.(*Server).handleStream
          | 	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1405
          | google.golang.org/grpc.(*Server).serveStreams.func1.1
          | 	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:746
          | runtime.goexit
          | 	/usr/local/go/src/runtime/asm_amd64.s:1374
        Wraps: (5) (n1,s1):6LEARNER: remote couldn't accept LEARNER_INITIAL snapshot 996e7fed at applied index 57 with error: [n1,s1],r36: cannot apply snapshot: [n1,s1]: canApplySnapshotLocked: cannot add placeholder, have an existing placeholder range=36 [/Table/Max-/Max) (placeholder) (n2,s2):5
        Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *markers.withMark (4) *errbase.opaqueWrapper (5) *errutil.leafError
E210120 19:30:15.134091 373380 sql/stats/automatic_stats.go:343  failed to get tables for automatic stats: get-tables: failed to send RPC: sending to all replicas failed; last error: rpc error: code = Canceled desc = grpc: the client connection is closing
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReportUnreachableRemoveRace142428455
--- FAIL: TestReportUnreachableRemoveRace (1.87s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestReportUnreachableRemoveRace PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestReportUnreachableRemoveRace.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -2520,7 +2520,6 @@ func TestReportUnreachableHeartbeats(t *testing.T) {
 // races (primarily in asynchronous coalesced heartbeats).
 func TestReportUnreachableRemoveRace(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 59209, "flaky test")
 	defer log.Scope(t).Close(t)
 
 	ctx := context.Background()
@@ -2529,9 +2528,9 @@ func TestReportUnreachableRemoveRace(t *testing.T) {
 			ReplicationMode: base.ReplicationManual,
 		})
 	defer tc.Stopper().Stop(ctx)
-
-	key := tc.ScratchRange(t)
+	key := tc.ScratchRangeWithExpirationLease(t)
 	tc.AddVotersOrFatal(t, key, tc.Targets(1, 2)...)
+	require.NoError(t, tc.WaitForVoters(key, tc.Targets(1, 2)...))
 
 outer:
 	for i := 0; i < 5; i++ {
@@ -2551,11 +2550,30 @@ outer:
 						tc.TransferRangeLeaseOrFatal(t, *repl.Desc(), tc.Target(replicaIdx))
 					}
 					tc.RemoveVotersOrFatal(t, key, tc.Target(leaderIdx))
-					cb := tc.Servers[replicaIdx].RaftTransport().GetCircuitBreaker(tc.Target(replicaIdx).NodeID, rpc.DefaultClass)
-					cb.Break()
+					// We want to stop all nodes from talking to the replicaIdx, so need
+					// to trip the breaker on all servers but it.
+					for i := range tc.Servers {
+						if i != replicaIdx {
+							cb := tc.Servers[i].RaftTransport().GetCircuitBreaker(tc.Target(replicaIdx).NodeID, rpc.DefaultClass)
+							cb.Break()
+						}
+					}
 					time.Sleep(tc.GetFirstStoreFromServer(t, replicaIdx).GetStoreConfig().CoalescedHeartbeatsInterval)
-					cb.Reset()
+					for i := range tc.Servers {
+						if i != replicaIdx {
+							cb := tc.Servers[i].RaftTransport().GetCircuitBreaker(tc.Target(replicaIdx).NodeID, rpc.DefaultClass)
+							cb.Reset()
+						}
+					}
+					// Make sure the old replica was actually removed, before we try to re-adding it.
+					testutils.SucceedsSoon(t, func() error {
+						if oldRepl := tc.GetFirstStoreFromServer(t, leaderIdx).LookupReplica(roachpb.RKey(key)); oldRepl != nil {
+							return errors.Errorf("Expected replica %s to be removed", oldRepl)
+						}
+						return nil
+					})
 					tc.AddVotersOrFatal(t, key, tc.Target(leaderIdx))
+					require.NoError(t, tc.WaitForVoters(key, tc.Target(leaderIdx)))
 					continue outer
 				}
 				t.Fatal("could not find raft replica")

